docs/advanced/automated-testing.md

```diff
@@ -0,0 +1,262 @@
+
+https://luarocks.org/
+&
+https://olivinelabs.com/busted/
+
+_Luarocks is known to be hard to impossible to install on Windows systems sucessfully, if that’s the case for you, definitely keep reading on, we’re gonna explore a (free) option to do away with running tests locally entirely._
+
+Now that we have busted installed and the `busted` command runs in your terminal of choice successfully we can proceed.
+
+Let’s quickly install busted typings like so `npm install -d busted-tstl`
+
+And we need to add busted-tstl types to our tsconfig.json, you should already have a section like this:
+
+```json title=tsconfig.json
+"types": [
+  "typescript-to-lua/language-extensions",
+  "lua-types/jit",
+]
+```
+
+Simply add `"busted-tstl"` :
+
+```json title=tsconfig.json
+"types": [
+  "typescript-to-lua/language-extensions",
+  "lua-types/jit",
+  "busted-tstl"
+]
+```
+
+Now let’s set up a folder for our tests, by default busted takes a folder named `spec` and runs the files in there, now
+this is personal preference, but usually I name my tests `[filename to be tested]_spec.ts`, the \_spec suffix is once again what busted searches for by default, so we’ll stick to that.
+
+Now our project should look something like this:
+
+![Project setup](/images/testing-project-setup-with-spec-folder.png)
+
+Alright, we can start writing our first test. Let’s explore the following example briefly:
+
+```typescript title=index_spec.ts
+import { libraryFunction } from "..";
+
+describe("Library Function", () => {
+  it("Returns unreversed string correctly", () => {
+    assert.is_equal("1,2,3,4", libraryFunction([1, 2, 3, 4], false));
+  });
+});
+```
+
+So the short version of what’s happening there is the ‘describe’ block is our named collection of tests, this collection tests the library function, so we've named it `"Library Function"`
+
+Next we call it(‘’,()=>{}) in there to name and describe the behaviour of a single test, it’s a regular function, so you can call and do anything in there. Here, we’re just using a function in the assert namespace which passes the test if the 2 parameters are exactly equal, and fails if they’re different.
+
+Alright, so we call our libraryFunction with a set of parameters, and get back some return, then we assert that the return value is equal to a known (correct) value. For illustration let’s add another test, since we have a second signature of the function that returns a reversed string, we can try that:
+
+```typescript title=index_spec.ts
+import { libraryFunction } from "..";
+
+describe("Library Function", () => {
+  it("Returns unreversed string correctly", () => {
+    assert.is_equal("1,2,3,4", libraryFunction([1, 2, 3, 4], false));
+  });
+  it("Returns reversed string correctly", () => {
+    assert.is_equal("4,3,2,1", libraryFunction([1, 2, 3, 4], true));
+  });
+});
+```
+
+Now hit compile and change your terminals directory in to your tstl output folder, for me it’s the `dist` folder.
```

Comment:
Would be nice to set this up as part of your TS workspace. To do this you can specify a custom 'test' script in package.json that builds and executes tests. An example of how I would set up my package.json in project root for your example repo:

```json
{
    ...
    "scripts": {
        ...
        "pretest": "tstl -p spec",
        "test": "busted spec",
        ...
    }
}
```

This would mean you no longer need to switch between directories. It also means you can build and execute your tests by simply running `npm test`, which is the standard way of running tests for node projects.

Have a look at the [TypeScriptToLua package.json](https://github.com/TypeScriptToLua/TypeScriptToLua/blob/master/package.json#L28-L29) as an example: